src/explorer/collectionTreeItem.ts

```diff
@@ -84,53 +108,100 @@ export default class CollectionTreeItem extends vscode.TreeItem
       return Promise.resolve([]);
     }
 
+    // Update cache if one of the children has been expanded/collapsed.
+    if (this.needsToUpdateCache()) {
+      this.cacheIsUpToDate = false;
+    }
+
+    if (this.cacheIsUpToDate) {
+      return Promise.resolve([
+        this._documentListChild,
+        this._schemaChild,
+        this._indexListChild
+      ]);
+    }
+
+    this.cacheIsUpToDate = true;
+
     // We rebuild the children here so their controlled `expanded` state
     // is ensure to be set by vscode.
     this._documentListChild = new DocumentListTreeItem(
       this.collectionName,
       this.databaseName,
       this._type,
       this._dataService,
-      this._documentListChild.isExpanded,
-      this._documentListChild.getMaxDocumentsToShow(),
-      this._documentListChild.getChildrenCache()
+      this._documentListChild ? this._documentListChild.isExpanded : false,
+      this._documentListChild
+        ? this._documentListChild.getMaxDocumentsToShow()
+        : MAX_DOCUMENTS_VISIBLE,
+      this._documentListChild
+        ? this._documentListChild.getChildrenCache()
+        : null
     );
+
     this._schemaChild = new SchemaTreeItem(
       this.collectionName,
       this.databaseName,
       this._dataService,
-      this._schemaChild.isExpanded,
+      this._schemaChild ? this._schemaChild.isExpanded : false,
       this._schemaChild
     );
-    return Promise.resolve([this._documentListChild, this._schemaChild]);
+
+    this._indexListChild = new IndexListTreeItem(
+      this.collectionName,
+      this.databaseName,
+      this._dataService,
+      this._indexListChild ? this._indexListChild.isExpanded : false,
+      this._indexListChild ? this._indexListChild.cacheIsUpToDate : false,
+      this._indexListChild ? this._indexListChild.getChildrenCache() : []
+    );
+
+    return Promise.resolve([
+      this._documentListChild,
+      this._schemaChild,
+      this._indexListChild
+    ]);
+  }
+
+  // eslint-disable-next-line complexity
+  needsToUpdateCache(): boolean {
+    return (
+      (this._documentListChild.isExpanded &&
```

Comment:
The complexity eslint is complaining about is kinda a real thing here. Is a bit hard to understand these conditions (especially for someone not having much context), and also probably quite easy to miscode them and introduce issues.
Could this code be factored further in a way that reveals the intent of the conditions? Just an idea: 

```
isChildCacheOutOfSync(child) {
   const isExpanded = child.isExpanded;
   const collapsibleState = child.collapsibleState;
   return isExpanded ? collapsibleState !== Expanded : collapsibleState !== Collapsed;
}
```

```
needsToUpdateCache() {
   isChildCacheOutOfSync(this._documentListChild) ||
   isChildCacheOutOfSync(this. _schemaChild) || ...
}
```

Reply:
Good suggestion - updated. ty